// @opentelemetry/otlp-grpc-exporter-base/build/src/index.js
"use strict";
/*
 * Copyright The OpenTelemetry Authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.LogsSerializer = exports.TraceSerializer = exports.MetricsSerializer = exports.validateAndNormalizeUrl = exports.DEFAULT_COLLECTOR_URL = exports.OTLPGRPCExporterNodeBase = void 0;
var OTLPGRPCExporterNodeBase_1 = require("./OTLPGRPCExporterNodeBase");
Object.defineProperty(exports, "OTLPGRPCExporterNodeBase", { enumerable: true, get: function () { return OTLPGRPCExporterNodeBase_1.OTLPGRPCExporterNodeBase; } });
var util_1 = require("./util");
Object.defineProperty(exports, "DEFAULT_COLLECTOR_URL", { enumerable: true, get: function () { return util_1.DEFAULT_COLLECTOR_URL; } });
Object.defineProperty(exports, "validateAndNormalizeUrl", { enumerable: true, get: function () { return util_1.validateAndNormalizeUrl; } });
var serializers_1 = require("./serializers");
Object.defineProperty(exports, "MetricsSerializer", { enumerable: true, get: function () { return serializers_1.MetricsSerializer; } });
Object.defineProperty(exports, "TraceSerializer", { enumerable: true, get: function () { return serializers_1.TraceSerializer; } });
Object.defineProperty(exports, "LogsSerializer", { enumerable: true, get: function () { return serializers_1.LogsSerializer; } });
